iroh-gossip/examples/chat.rs

```diff
@@ -0,0 +1,345 @@
+//! signatures on received messages.
+//!
+//! To connect to a swarm, you need to share a DERP server and know at least
+//! one other peer's peer id.
+//!
+//! By default a new peer id is created when starting the example. To reuse your identity,
+//! set the `--private-key` CLI flag with the private key printed on a previous invocation.
+//!
+//! You can use this with a local DERP server. To do so, run
+//! `cargo run --bin derper -- --dev`
+//! and then set the `-d http://localhost:3340` flag on this example.
+
+use std::{collections::HashMap, fmt, net::SocketAddr, str::FromStr};
+
+use anyhow::bail;
+use bytes::Bytes;
+use clap::Parser;
+use ed25519_dalek::{Signature, Signer, SigningKey, VerifyingKey};
+use iroh_gossip::{
+    net::{GossipHandle, GOSSIP_ALPN},
+    proto::{Event, TopicId},
+};
+use iroh_net::{
+    defaults::{default_derp_map, DEFAULT_DERP_STUN_PORT},
+    hp::derp::{DerpMap, UseIpv4, UseIpv6},
+    magic_endpoint::accept_conn,
+    tls::{Keypair, PeerId},
+    MagicEndpoint,
+};
+use once_cell::sync::OnceCell;
+use serde::{Deserialize, Serialize};
+use tokio::sync::mpsc;
+use url::Url;
+
+#[derive(Parser, Debug)]
+struct Args {
+    /// Private key to derive our peer id from
+    #[clap(long)]
+    private_key: Option<String>,
+    /// Set a custom DERP server. By default, the DERP server hosted by n0 will be used.
+    #[clap(short, long)]
+    derp: Option<Url>,
+    /// Disable DERP completeley
+    #[clap(long)]
+    no_derp: bool,
+    /// Set your nickname
+    #[clap(short, long)]
+    name: Option<String>,
+    /// Set the bind port for our socket. By default, a random port will be used.
+    #[clap(short, long, default_value = "0")]
+    bind_port: u16,
+    #[clap(subcommand)]
+    command: Command,
+}
+
+#[derive(Parser, Debug)]
+enum Command {
+    OpenTopic { topic: String },
+    JoinTicket { ticket: String },
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    tracing_subscriber::fmt::init();
+    let args = Args::parse();
+
+    // parse or generate our keypair
+    let keypair = match args.private_key {
+        None => Keypair::generate(),
+        Some(key) => parse_keypair(&key)?,
+    };
+    println!("> our private key: {}", fmt_secret(&keypair));
+
+    // configure our derp map
+    let derp_map = match (args.no_derp, args.derp) {
+        (false, None) => Some(default_derp_map()),
+        (false, Some(url)) => Some(derp_map_from_url(url)?),
+        (true, None) => None,
+        (true, Some(_)) => bail!("You cannot set --no-derp and --derp at the same time"),
+    };
+    println!("> using DERP servers: {:?}", fmt_derp_map(&derp_map));
+
+    // init a cell that will hold our gossip handle to be used in endpoint callbacks
+    let gossip_cell: OnceCell<GossipHandle> = OnceCell::new();
+    // init a channel that will emit once the initial endpoints of our local node are discovered
+    let (initial_endpoints_tx, mut initial_endpoints_rx) = mpsc::channel(1);
```

Comment:
it looks like this should be a oneshot? (`tokio::sync::oneshot`)

Reply:
Conceptually it should be, yes! However to use tokio's oneshot channel we'd have to wrap it in a `Arc<Mutex<Option>>>` unfortunately, which I thought to be a bit much for this example. Added this comment to explain in the code:
```
    // (not using a oneshot channel here because tokio::oneshot::Sender::send takes self, which
    // does not work in the on_endpoints Fn closure)
```